torch/fx/experimental/sym_node.py

```diff
@@ -755,8 +755,83 @@ def _sympy_rshift(a, b):
     return RShift(a, b)
 
 
+def _binary_search_insert_arg(ordered_args: List[sympy.Basic], new_arg: sympy.Basic):
+    """
+    If new_arg is found in ordered_args None is returned, else the new
+    ordered_args with new_arg inserted
+    """
+    low, high = 0, len(ordered_args) - 1
+    from sympy.core.basic import _args_sortkey as sort_key
+
+    # Fast path when new_arg > ordered_args[-1], which is common.
+    if len(ordered_args) and sort_key(ordered_args[-1]) < sort_key(new_arg):
+        return ordered_args + [new_arg]
+
+    while low <= high:
+        mid = (low + high) // 2
+        if ordered_args[mid] == new_arg:
+            return None
+        elif sort_key(ordered_args[mid]) < sort_key(new_arg):
+            low = mid + 1
+        else:
+            high = mid - 1
+
+    ordered_args.insert(low, new_arg)
```

Comment:
Doesn't this make this function O(n) anyways though??

Reply:
oh yeh good catch but this is saves bunch of seconds it reduces the comparisons. 
just ran it again on 1000 nodes. and confirmed wins evey with recent sympy changes. 
maybe the add does more than just the sorting?

Reply:
one thing is that comparing symbols is expensive, I know the recent sympy changes optimizes that by caching stuff.

Reply:
here we go that answers it 
<img width="1448" alt="Screenshot 2024-11-13 at 11 06 57 PM" src="https://github.com/user-attachments/assets/4e79860b-0021-4db2-8cd7-70750e741afd">

the profile of constructing Add

Reply:
sg